src/libecalc/core/models/compressor/train/variable_speed_compressor_train_common_shaft_multiple_streams_and_pressures.py

```diff
@@ -405,6 +407,70 @@ def get_max_standard_rate_per_stream(
+        can be added.
+
+        """
+        stream_to_maximize_connected_to_stage_no = self.streams[stream_to_maximize].connected_to_stage_no
+
+        # if it is not an ingoing stream --> currently no calculations done
+        # Fixme: what should be returned? 0.0, NaN or something else?
+        if not self.streams[stream_to_maximize].is_inlet_stream:
+            return 0.0
+
+        std_rates_std_m3_per_day_per_stream = rate_per_stream.copy()
+
+        def _calculate_train_result(std_rate_for_stream: float) -> CompressorTrainResultSingleTimeStep:
+            """Partial function of self.calculate_compressor_train_given_rate_ps_speed
+            where we only pass std_rate_per_stream and speed.
+            """
+            std_rates_std_m3_per_day_per_stream[stream_to_maximize] = std_rate_for_stream
+            return self.calculate_shaft_speed_given_rate_ps_pd(
+                std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
+
+        train_result = self.calculate_shaft_speed_given_rate_ps_pd(
+            std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+            suction_pressure=suction_pressure,
+            target_discharge_pressure=target_discharge_pressure,
+        )
+        if not train_result.is_valid:
+            zero_stream_result = _calculate_train_result(std_rate_for_stream=EPSILON)
+            if not zero_stream_result.is_valid:
+                return 0.0
+            else:
+                return maximize_x_given_boolean_condition_function(
+                    x_min=0.0,
+                    x_max=float(rate_per_stream[stream_to_maximize]),
+                    bool_func=lambda x: _calculate_train_result(std_rate_for_stream=x).is_valid,
+                )
+        else:
+            max_rate_is_larger_than = std_rates_std_m3_per_day_per_stream[stream_to_maximize]
+        while train_result.is_valid:
+            max_rate_is_larger_than = train_result.stage_results[
+                stream_to_maximize_connected_to_stage_no
+            ].standard_rate_asv_corrected_sm3_per_day
+            std_rates_std_m3_per_day_per_stream[stream_to_maximize] = max_rate_is_larger_than * 2
+            train_result = self.calculate_shaft_speed_given_rate_ps_pd(
+                std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
+        return maximize_x_given_boolean_condition_function(
+            x_min=float(max_rate_is_larger_than),
+            x_max=float(max_rate_is_larger_than * 2),
+            bool_func=lambda x: _calculate_train_result(std_rate_for_stream=x).is_valid,
+        )
```

Comment:
And this find a more accurate max rate?

Reply:
And here we find the point where it switches trom true to false

Reply:
Among the "points we have" or as soon as we are within EPSILON of "where it switches"?